tests-e2e/gappy_state_test.go

```diff
@@ -0,0 +1,117 @@
+package syncv3_test
+
+import (
+	"fmt"
+	"github.com/matrix-org/sliding-sync/sync3"
+	"github.com/matrix-org/sliding-sync/testutils/m"
+	"github.com/tidwall/gjson"
+	"testing"
+)
+
+// Test that state changes "missed" by a poller are injected back into the room when a
+// future poller recieves a v2 incremental sync with a state block.
+func TestGappyState(t *testing.T) {
+	t.Log("Alice registers on the homeserver.")
+	alice := registerNewUser(t)
+
+	t.Log("Alice creates a room")
+	firstRoomName := "Romeo Oscar Oscar Mike"
+	roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": firstRoomName})
+
+	t.Log("Alice sends a message into that room")
+	firstMessageID := alice.SendEventSynced(t, roomID, Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Hello, world!",
+		},
+	})
+
+	t.Log("Alice requests an initial sliding sync on device 1.")
+	syncResp := alice.SlidingSync(t,
+		sync3.Request{
+			Lists: map[string]sync3.RequestList{
+				"a": {
+					Ranges: [][2]int64{{0, 20}},
+					RoomSubscription: sync3.RoomSubscription{
+						TimelineLimit: 10,
+					},
+				},
+			},
+		},
+	)
+	m.MatchResponse(
+		t,
+		syncResp,
+		m.MatchRoomSubscription(
+			roomID,
+			m.MatchRoomName(firstRoomName),
+			func(r sync3.Room) error {
+				if len(r.Timeline) == 0 {
+					return fmt.Errorf("no/empty timeline in response")
+				}
+				lastReceivedEventID := gjson.ParseBytes(r.Timeline[len(r.Timeline)-1]).Get("event_id").Str
```

Comment:
You should be able to use `MatchRoomTimelineMostRecent(1, Event{ ID: firstMessageID })` to get this behaviour, so you can delete this function. Note that these are _not_ in the `m.` package because the test doesn't know the full event JSON.